api/services/bulk-docs.js

```diff
@@ -35,35 +34,69 @@ const generateBatches = (docs, batchSize) => {
   return batches;
 };
 
-const generateBatchPromise = (batch, res, { isFinal } = {}) => {
-  return db.medic.bulkDocs(batch).then(result => {
-    let resString = JSON.stringify(result);
-    resString += isFinal ? '' : ',';
-    res.write(resString);
-  });
+const getDocsToModify = ids => {
+  let docsToDelete;
+  return db.medic.allDocs({ keys: ids, include_docs: true })
+    .then(result => {
+      docsToDelete = markAsDeleted(result);
+      return utils.updateParentContacts(docsToDelete);
+    })
+    .then(docsToUpdate => {
+      const allDocs = docsToDelete.concat(docsToUpdate);
+      checkForDuplicates(allDocs);
+      return allDocs;
+    });
 };
 
-module.exports = {
-  bulkDelete: (docs, res, { batchSize = 100 } = {}) => {
-    let docsToDelete;
-    const keys = docs.map(doc => doc._id);
-    return db.medic.allDocs({ keys, include_docs: true })
-      .then(result => {
-        docsToDelete = markAsDeleted(result);
-        return utils.updateParentContacts(docsToDelete);
+const deleteBatches = (batches, res, docsToRetry, attemptNumber) => {
+  return batches.reduce((promise, batch, index) => {
+    const isFinal = index === batches.length - 1;
+    let docsToModify;
+    return promise
+      .then(() => {
+        return getDocsToModify(batch);
+      })
+      .then(allDocs => {
+        docsToModify = allDocs;
+        return db.medic.bulkDocs(docsToModify);
       })
-      .then(docsToUpdate => {
-        const allDocs = docsToDelete.concat(docsToUpdate);
-        checkForDuplicates(allDocs);
-        const batches = generateBatches(allDocs, batchSize);
+      .then(result => {
+        const errors = result.map(docUpdate => docUpdate.error);
```

Comment:
Here you're deciding something has errored if there is an `error` property, and below you're checking if something is successful if there is a truthy `ok` property. Maybe these checks should use the same property for clarity?

Also, should you be ignoring 404s here?